### ISSUE TITLE
multiprocessing race condition

### ISSUE THREAD

**Issue by lincoln-harris** [czbiohub/cerebra#47]
was getting a biopython.seq error when running through some new vcf files:

`TranslationError: Codon '>' is invalid`

but seems to have resolved its self somehow

**Comment by lincoln-harris** [comment]
full traceback
```Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 341, in __getitem__
    self.ambiguous_nucleotide)
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 193, in list_possible_proteins
    x1 = ambiguous_nucleotide_values[c1]
KeyError: '>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Seq.py", line 2107, in _translate_str
    amino_acids.append(forward_table[codon])
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 344, in __getitem__
    raise KeyError(codon)  # stop codon
KeyError: '>GL'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/multiprocess/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/Users/lincoln.harris/code/cerebra/cerebra/find_aa_mutations.py", line 215, in process_cell
    .find_cell_gene_aa_mutations(path=path)
  File "/Users/lincoln.harris/code/cerebra/cerebra/find_aa_mutations.py", line 149, in find_cell_gene_aa_mutations
    .predict_for_vcf_record(record)
  File "/Users/lincoln.harris/code/cerebra/cerebra/protein_variant_predictor.py", line 259, in predict_for_vcf_record
    ref_aa_seq = ref_coding_seq.translate()
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Seq.py", line 1038, in translate
    cds, gap=gap)
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/Bio/Seq.py", line 2124, in _translate_str
    "Codon '{0}' is invalid".format(codon))
Bio.Data.CodonTable.TranslationError: Codon '>GL' is invalid
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/cerebra/bin/cerebra", line 11, in <module>
    load_entry_point('cerebra', 'console_scripts', 'cerebra')()
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/lincoln.harris/code/cerebra/cerebra/find_aa_mutations.py", line 285, in find_aa_mutations
    processes=num_processes)
  File "/Users/lincoln.harris/code/cerebra/cerebra/find_aa_mutations.py", line 224, in find_transcript_mutations
    smoothing=0.01))
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/tqdm/std.py", line 1087, in __iter__
    for obj in iterable:
  File "/usr/local/anaconda3/envs/cerebra/lib/python3.6/site-packages/multiprocess/pool.py", line 735, in next
    raise value
Bio.Data.CodonTable.TranslationError: Codon '>GL' is invalid```

**Comment by lincoln-harris** [comment]
also I'm only seeing this when I run with > 1 processes....perhaps we're not handling `Bio.Seq` warnings correctly? 
_eg._ 
```
BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  BiopythonWarning)

**Comment by rvanheusden** [comment]
Looks like this issue is occurring here:
https://github.com/czbiohub/cerebra/blob/071c366a4dc4fb3f19f8de5b2585e2df785c8ee6/cerebra/protein_variant_predictor.py#L264

In this case, it seems to be trying to translate `>GL`, which should be derived from the reference sequence. If `>GL` is appearing in the reference sequence, that might indicate that something is off. `>GL` looks like it could even be the start of a FASTQ sequence identifier? I would check out the reference sequence file to see if you can find those characters. Just my two cents from a cursory look.

**Comment by lincoln-harris** [comment]
thanks rohan. yeah for some reason its trying to translate a fasta header. I'm trying to isolate the aberrant files and figure out whats different about them

**Comment by lincoln-harris** [comment]
only seeing when i run with processes > 1? that doesnt make any sense

**Comment by lincoln-harris** [comment]
behavior is stochastic -- 'problematic' files are not the same from test to test

**Comment by rvanheusden** [comment]
I have a theory...

The protein variant predictor expects FASTA sequences to be provided through a `pyfaidx` object. The fasta index should be loaded before any additional processes are spawned, but perhaps the sequence fetching behavior is susceptible to issues when multiple processes attempt to read from the index simultaneously.

**Comment by lincoln-harris** [comment]
`pyfaidx Fasta` in theory is thread-safe
https://github.com/mdshw5/pyfaidx/issues/92

**Comment by lincoln-harris** [comment]
i THINK this is resolved by PR #50

**Comment by lincoln-harris** [comment]
RESOLVED, by #56
